New Insights config API proposal (#1037)
openshift/enhancements: enhancements/insights/insights-config-api.md

## Motivation

The main motivation for this new API is to allow users to refresh (or force) the gathering of the Insights data (see the [User stories](#user-stories) section for more details) as well as to disable the gathering completely. This new API will allows us to report various Insights status reports (including the Insights analysis report that is now exposed via Prometheus metrics).

### Goals

The goal of this enhancement is to propose a basic structure for a new Insights configuration API and define new configuration options. This includes following:

- Define a new configuration option providing a way to trigger on demand Insights gathering
- Define a new configuration option providing a way to include/exclude particular Insights data gatherers
- Provide Insights status & reporting in the status field in the new API

### Non-Goals

- Move the majority of the current configuration options to a new configuration API - Insights has quite a lot of configuration options
  and it does not make sense to expose everything in the new configuration API 

> - Removing the original configuration via `support` secret

Reviewer: Any particular reason why not? What about the precedence? Ie. what is more important the `support` secret or the API configuration if they are touching similar areas?

Author: This is described in the https://github.com/openshift/enhancements/pull/1037/files#diff-e44167fc640994cf3d2a0afef771e4f11d3be51ccf7e5209c1cd8c9c81e054a1R260 and in the https://github.com/openshift/enhancements/pull/1037/files#diff-e44167fc640994cf3d2a0afef771e4f11d3be51ccf7e5209c1cd8c9c81e054a1R268

Reviewer: I'd like this `"support" secret config takes precedence over the new config API` to be explicitly mentioned right below after you introduce this new API. Test plan is fine, but that's not where I look for key assumptions.

Author: I added it to https://github.com/openshift/enhancements/pull/1037/files#diff-e44167fc640994cf3d2a0afef771e4f11d3be51ccf7e5209c1cd8c9c81e054a1R117 ...not sure if it's the best place....

Reviewer: I'd suggest moving it at the top of the Proposal section, it's lost between API types.

Author: OK. Moved.

## Proposal

The proposal is to introduce new Insights `CustomResourceDefinition` types in the OpenShift API providing a way to configure the Insights operator and to report its status. 

The proposal for the new `insights.config.openshift.io` resource is:

```go

> type InsightsSpec struct {

Author: Added

	// GatheringConfig spec attribute includes all the configuration options related to
	// gathering of the Insights archive and its uploading to the ingress.
	GatheringConfig *GatheringConfig `json:"gatheringConfig,omitempty"`
}

type InsightsStatus struct {
}

type GatheringConfig struct {
	// DataPolicy allows user to enable additional global obfuscation of the IP addresses and base domain
	// in the Insights archive data.
	// +kubebuilder:default=NoPolicy
	DataPolicy DataPolicy `json:"dataPolicy"`
	// ForceGatherReason enables user to force Insights data gathering by setting a new reason.
	// When there is some gathering in the progress then it is interrupted.
	// When all the gatherers are deactivated by the `DisabledGatherers`, nothing happens.

> 	ForceGatherReason string `json:"forceGatherReason"`

Author: The value will be cleared as we discussed yesterday. I added the comment

	// List of gatherers to be excluded from the gathering. All the gatherers can disabled by providing "all" value.
	// If all the gatherers are disabled, the Insights operator does not gather any data.
	// The particular gatherers IDs can be found at https://github.com/openshift/insights-operator/blob/master/docs/gathered-data.md.
	// An example of disabling some gatherers can look like `disabledGatherers: ["clusterconfig/machine_configs", "workloads/workload_info"]`
	DisabledGatherers []string `json:"disabledGatherers"`
}

const (
	// No data obfuscation
	NoPolicy DataPolicy = "NoPolicy"
	// IP addresses and cluster domain name is obfuscated
	IPsAndClusterDomainPolicy DataPolicy = "IPsAndClusterDomainPolicy"
)

// DataPolicy is a valid value for GatheringConfig.DataPolicy
type DataPolicy string
```

The proposal for the new `insights.operator.openshift.io` resource is:

```go

> type InsightsSpec struct {

Author: Added

	OperatorSpec `json:",inline"`
}

type InsightsStatus struct {
	// GatheringStatus provides basic information about the last Insights gathering.
	GatheringStatus *GatheringStatus `json:"gatheringStatus,omitempty"`
	// ReportStatus provides general Insights analysis results.
	ReportStatus *ReportStatus `json:"reportStatus,omitempty"`
}

type GatheringStatus struct {
	// LastGatherTime is the last time when Insights gathering finished.
	LastGatherTime metav1.Time `json:"lastGatheringTime,omitempty"`
	// Reason provides last known reason of gathering. This is helpful
	// especially when gathering was forced by user
	Reason string `json:"reason,omitempty"`
	// Status is waiting or running. If it is running then it means there is some gathering in progress.
	// If it is waiting then a new gathering can be triggered immediately.
	Status *GatherStatus `json:"status,omitempty"`
	// List of active gatherers (and their statuses) in the last gathering.
	GathererStatuses []GathererStatus `json:"gathererStatuses,omitempty"`

> }

Author: Thanks. Updated as suggested.

type ReportStatus struct {
	// Number of active Insights healthchecks with low severity
	LowCount int `json:"low"`
	// Number of active Insights healthchecks with moderate severity
	ModerateCount int `json:"moderate"`
	// Number of active Insights healthchecks with important severity
	ImportantCount int `json:"important"`
	// Number of active Insights healthchecks with critical severity
	CriticalCount int `json:"critical"`
	// TotalCount is the count of all active Insights healthchecks

> 	TotalCount int `json:"total"`

Author: Updated

}

type GathererStatus struct {
	// Name is the name of the gatherer.
	Name string `json:"name"`
	// GathererConditions provide details on the status of each gatherer.
	GathererConditions []GathererCondition `json:"conditions"`
	// DurationMillisecond represents the time spent gathering.
	DurationMillisecond int64 `json:"durationMillisecond"`
	// RecordsCount represents the number of records produced by the gatherer.

> 	RecordsCount int `json:"recordsCount"`

Author: It's an actual number of files/records created by the particular gatherer in the Insights archive. I guess I should extend the comment right?

}

type GathererCondition struct {
	// Type of the gatherer condition
	Type GathererConditionType `json:"type"`
	// Status is last known status of the particular gatherer condition
	Status corev1.ConditionStatus `json:"status"`
	// Messages is an optional attribute that provides error and warning messages
	// from the gatherer

> 	Messages []string `json:"messages,omitempty"`

Author: Added.

> 	Messages []string `json:"messages,omitempty"`

Reviewer: I didn't notice that earlier, but since each gatherer has its own status `GathererStatus`, and within that you have an array of conditions `GathererConditions` you could probably change this Messages array to be just `Message string` I don't expect that much of data in it, based on my previous experience.

> 	Messages []string `json:"messages,omitempty"`

Author: This is the same as the https://github.com/openshift/enhancements/pull/1037#discussion_r844961191. I think it's more convenient as a slice for our purposes. We can have more various messages as in https://github.com/openshift/insights-operator/blob/master/docs/insights-archive-sample/insights-operator/gathers.json#L7

}

// GathererConditionType is a valid value for GathererCondition.Type
type GathererConditionType string

const (
	// Gatherer was successful withnout any error reported
	Successful GathererConditionType = "Successful"
	// Gatherer was disabled by user
	Disabled GathererConditionType = "Disabled"
	// Gatherer failed to run
	Failed GathererConditionType = "Failed"
	// Gatherer was running, but there were some errors
	Warning GathererConditionType = "Warning"
	// Gatherer panicked

> 	Panic GathererConditionType = "Panic"

Author: Removed the panic type

)

const (
	// There is no gathering in progress
	GatherStatusWaiting GatherStatus = "Waiting"
	// There is running gathering in progress
	GatherStatusRunning GatherStatus = "Running"

> )

Author: Removed.

// GatherStatus is a valid value for GatheringStatus.Status
type GatherStatus string
```

### User Stories
#### Force gathering - refresh Insights status
As an OCP user/cluster-admin I would like to be able to trigger on demand gathering (there were some active Insights health checks in the cluster, which should be fixed and I want to refresh the Insights status). Steps:
- the user will set a new value of `ForceGatheringReason`
- if all the gatherers are disabled (`disabledGatherers: ["all"]`), nothing happens
- if there is some gathering in progress (`Status` is "running" in the `GatheringStatus` section) then it will be interrupted and the new gathering will be triggered
- if there is not any gathering in progress (`Status` is "waiting" in the `GatheringStatus` section) then the new gathering will be triggered almost immediately

> - user can check the `Status` or the `LastGatheringTime` attribute in the `GatheringStatus` section to check if the new gathering already finished

Reviewer: Same question as above, what happens with `ForceGatheringReason` field after a successful run?

Reviewer: I compared this with `ForceRedeploymentReason` in our operators, and they leave it set, so for consistency I'm fine leaving it set. You'll easily know when the field changes to initiate a new run from informers.

#### Disable Insights gathering completely
As an OCP user/cluster-admin I would like to disable Insights gathering completely. Steps:
- Set the `DisabledGatherers` to `["all"]` and the gathering will be turned off - i.e no new data will be gathered

#### Exposing latest Insights status
As an OCP user/cluster-admin I would like to see the latest Insights status. The status is represented by the `InsightsStatus` and it is only updated by the operator.

#### Excluding particular gatherers
As an OCP user/cluster-admin I would like to exclude particular gatherers. Steps:
- Set the `DisabledGatherers` with the list of gatherers to be excluded/disabled - e.g `DisabledGatherers: ["clusterconfig/config_maps", "clusterconfig/node_logs"]` 

### API Extensions

This is adding new `insights.config.openshift.io` and `insights.operator.openshift.io` API resources. The proposal of the structure of the new API is mentioned in the [Proposal](#Proposal) section.

### Risks and Mitigations

Risk: Insights operator is not able to read the new config API.

> Mitigation: The Insights operator is marked as unvailable through the corresponding clusteroperator condition.

Reviewer: You mean degraded?

Author: OK updated to degraded

---
title: insights-config-api
authors:
  - "@tremes"
reviewers:
  - "@bparees"
  - "@deads2k"
  - "@mfojtik"
  - "@wking"
  - "@soltysh"
approvers:
  - "@bparees"
  - "@deads2k"
  - "@mfojtik"
api-approvers:
  - "@deads2k"

>   - "@mfojtik"

Reviewer: Add me to both of these too.

Author: Done

	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata"`

	// spec is the specification of the desired behavior of the Insights
	// +kubebuilder:validation:Required
	// +required
	Spec InsightsSpec `json:"spec"`

	// status is the most recently observed status of the Insights operator
	// +optional
	Status InsightsStatus `json:"status"`
}

type InsightsSpec struct {
	OperatorSpec `json:",inline"`
}

type InsightsStatus struct {
	// GatheringStatus provides basic information about the last Insights gathering.
	GatheringStatus *GatheringStatus `json:"gatheringStatus,omitempty"`
	// ReportStatus provides general Insights analysis results.
	ReportStatus *ReportStatus `json:"reportStatus,omitempty"`
}

type GatheringStatus struct {
	// LastGatherTime is the last time when Insights gathering finished.
	LastGatherTime metav1.Time `json:"lastGatherTime,omitempty"`
	// LastGatherReason provides last known reason of gathering. This is helpful
	// especially when gathering was forced by user
	LastGatherReason string `json:"lastGatherReason,omitempty"`
	// StartGatherTime is the time when gathering started. The value is nil
	// when there is no gathering in progress.

> 	StartGatherTime metav1.Time `json:"startGatherTime,omitempty"`

Reviewer: If you want to make it nil, you need to make it a pointer and make it `// +nullable`, iow:
```golang
// +nullable
StartGatherTime *metav1.Time `json:"startGatherTime,omitempty"`
```

Reviewer: Alternatively you can change the description to say it will be 0, metav1 has `IsZero` function check, and that's fine as well. Just make it clear, because the current wording is confusing.

Author: Updated with 0 as default value

	// Reason for the condition's last transition.
	Reason string `json:"reason,omitempty" protobuf:"bytes,5,opt,name=reason"`
}

// GathererConditionType is a valid value for GathererCondition.Type
type GathererConditionType string

const (
	// Gatherer was successful withnout any error reported
	Successful GathererConditionType = "Successful"
	// Gatherer was disabled by user
	Disabled GathererConditionType = "Disabled"
	// Gatherer failed to run
	Failed GathererConditionType = "Failed"
	// Gatherer was running, but there were some errors
	Warning GathererConditionType = "Warning"
)
```

### User Stories
#### Force gathering - refresh Insights status
As an OCP user/cluster-admin I would like to be able to trigger on demand gathering (there were some active Insights health checks in the cluster, which should be fixed and I want to refresh the Insights status). Steps:
- the user will set a new value of `ForceGatherReason`
- if all the gatherers are disabled (`disabledGatherers: ["all"]`), nothing happens
- if there is some gathering in progress (`StartGatherTime` is not nil in the `GatheringStatus` section) then it will be interrupted and the new gathering will be triggered

> - if there is not any gathering in progress (`StartGatherTime` is nil in the `GatheringStatus` section) then the new gathering will be triggered almost immediately

Reviewer: ```suggestion
- if there is no gathering in progress (`StartGatherTime` is nil in the `GatheringStatus` section) then the new gathering will be triggered almost immediately
```

Author: Thanks

- user can check the `StartGatherTime` or the `LastGatheringTime` attribute in the `GatheringStatus` section to check if the new gathering already finished
- when the gathering is finished then the value of `ForceGatherReason` is cleared and the `LastGatherReason` and `LastGatherTime` in the `GatheringStatus` are updated

#### Disable Insights gathering completely
As an OCP user/cluster-admin I would like to disable Insights gathering completely. Steps:
- Set the `DisabledGatherers` to `["all"]` and the gathering will be turned off - i.e no new data will be gathered

#### Exposing latest Insights status
As an OCP user/cluster-admin I would like to see the latest Insights status. The status is represented by the `InsightsStatus` and it is only updated by the operator.

#### Excluding particular gatherers
As an OCP user/cluster-admin I would like to exclude particular gatherers. Steps:
- Set the `DisabledGatherers` with the list of gatherers to be excluded/disabled - e.g `DisabledGatherers: ["clusterconfig/config_maps", "clusterconfig/node_logs"]` 

### API Extensions

> This is adding new `insights.config.openshift.io` and `insights.operator.openshift.io` API resources. The proposal of the structure of the new API is mentioned in the [Proposal](#Proposal) section.

Reviewer: To be precise the full name for the newly added resources should be:
- Insights in config.openshift.io/v1 API group
- Insights in operator.openshift.io/v1 API group 
:wink:

Author: Thanks. Updated.

### Risks and Mitigations

Risk: Insights operator is not able to read the new config API.

Mitigation: The Insights operator is marked as degraded in the corresponding clusteroperator condition.

Risk: Somebody/something is repeatedly updating the `ForceGatherReason` attribute and thus forcing gathering. There is a risk that 
the cluster would upload data too often, but this assumes that the "repeating interval" allows uploading (if it's forced too quickly then there will be no upload, because the gathering is not over).

Mitigation: Maybe we should introduce some limit on the number of forced gatherings within 1 hour, for example. 

## Design Details

### On demand data gathering

The on demand data gathering can happen almost immediately when there is no active gathering in the progress, but when there is some in progress 
then it will be interrupted and the new gathering will be triggered. 

The forced gathering is defined as follows: 
- force it - e.g using a `ForceGatherReason` attribute where the user would provide some string as reason (mentioned in the [Proposal](#proposal)). 
  - pros

>     - [providing a reason (e.g “Refresh Insights”) is simpler for configuration than providing a time

Reviewer: ```suggestion
    - providing a reason (e.g “Refresh Insights”) is simpler for configuration than providing a time
```

#### Dev Preview -> Tech Preview

The original way of configuring the Insights is still available. If there is a conflict, in other words, if there is some configuration option defined in the `support` secret as well as in the new configuration API then the value from the `support` secret takes precedence. 

#### Tech Preview -> GA

The `TechPreviewNoUpgrade` feature gate requirement is removed. The behavior defined above in the [Dev Preview -> Tech Preview](#dev-preview---tech-preview) section is still true. 

#### Removing a deprecated feature

The existing configuration via `support` secret should be still available for backward compatibility reasons. The reason of the `support` secret existence is the option of using basic authentication for the Ingress service, which Insights operator uses as the upload endpoint. 

### Upgrade / Downgrade Strategy

No update of the configuration is required. The existing configuration in the `support` secret continues to work. When there is an existing `insights.config.openshift.io` kind in a cluster after upgrade, then the user can optionally use the new configuration API. 

> There can be a conflict when a configuration option is defined in the `support` secret as well as in the configuration API. The configuration in the `support` secret takes precedence in this case. 

Reviewer: This precedence will also handle the downgrade case, when you move from a cluster with that Insigths API to one without it, the fact you will pick `support` over API is to your benefit. The only downside is when API disappears but the `support` ticket wasn't present. Please add both of these here, for clarity

### Version Skew Strategy

### Operational Aspects of API Extensions

This API is provided exclusively for the Insights operator and there will be no other component depending on this new API. Current configobserver
in the Insights operator runs every 5 minutes so if we preserve the same behavior then there will be a request to the API every 5 minutes. Additionally the field

> for the forced gathering (`ForceGatherReason` - see [Proposal](#proposal)) will require an informer.

Reviewer: I wouldn't expect API call every 5 mins, you'll open a watch (using informer) which will notify you of all the changes.

Author: Updated

	metav1.ObjectMeta `json:"metadata,omitempty"`

	// spec holds user settable values for configuration
	// +kubebuilder:validation:Required
	// +required
	Spec InsightsSpec `json:"spec"`
	// status holds observed values from the cluster. They may not be overridden.
	// +optional
	Status InsightsStatus `json:"status"`
}

type InsightsSpec struct {
	// GatheringConfig spec attribute includes all the configuration options related to
	// gathering of the Insights archive and its uploading to the ingress.
	GatheringConfig *GatheringConfig `json:"gatheringConfig,omitempty"`
}

type InsightsStatus struct {
}

type GatheringConfig struct {
	// DataPolicy allows user to enable additional global obfuscation of the IP addresses and base domain
	// in the Insights archive data.
	// +kubebuilder:default=NoPolicy
	DataPolicy DataPolicy `json:"dataPolicy"`
	// ForceGatherReason enables user to force Insights data gathering by setting a new reason.
	// When there is some gathering in the progress then it is interrupted.
	// When all the gatherers are deactivated by the `DisabledGatherers`, nothing happens.
	// When the forced gathering is finished then the value is cleared.

> 	ForceGatherReason string `json:"forceGatherReason"`

Reviewer 2: This seems transient and the rest seem permanent.  I'd make two separate values under spect for them.

Reviewer: @deads2k are you referring to forced gathering (transient) and the disablement of gatherers (permanent)? If so, that's already handled separately. This field is only for the former, the latter is done through `DisabledGatherers` below. Or are you talking about something else?

Reviewer 2: discussed in slack.  Also, this could get worked out in an API PR.  It's important, but the fitting impact here is relatively minor

	// List of gatherers to be excluded from the gathering. All the gatherers can be disabled by providing "all" value.
	// If all the gatherers are disabled, the Insights operator does not gather any data.
	// The particular gatherers IDs can be found at https://github.com/openshift/insights-operator/blob/master/docs/gathered-data.md.
	// An example of disabling gatherers looks like this: `disabledGatherers: ["clusterconfig/machine_configs", "workloads/workload_info"]`
	DisabledGatherers []string `json:"disabledGatherers"`
}

const (
	// No data obfuscation
	NoPolicy DataPolicy = "NoPolicy"
	// IP addresses and cluster domain name is obfuscated
	IPsAndClusterDomainPolicy DataPolicy = "IPsAndClusterDomainPolicy"
)

type DataPolicy string
```

> Note that the `support` secret configuration still takes precedence over the new config API. 

Reviewer 2: This seems backwards.  If someone is using this, the old should be retired and perhaps force-assigned to reflect these values.

Reviewer 3: >  If someone is using this, the old should be retired and perhaps force-assigned to reflect these values.

i might not be following what "force reassigned to reflect these values means" (forced in which direction), but fundamentally if someone has config in a secret, that config should take precedence over any default values implied by the introduction of this new api resorce with empty values, no?

and if they have existing tooling that's ignorant of this new api and continues to update the secret, i'd expect those config values to be respected, until such time as we actually remove support for the "secret-based" config api (which, being that it's an api in my mind, we can't do without some sort of deprecation process)

Author: Another point is that the current proposal doesn't provide all the config options available via the `support` secret so we can't simply deprecate it IMO.

Reviewer: @deads2k based on my understanding after talking with @tremes there is no clear goal yet, to fully replace these config secret, that's why the secret precedence here makes sense. Until there's a clear desire to move out of the config we can switch the priorities, but for now, given only limited set of functionality is supported there I'm ok with this approach.

The proposal for the new `insights.operator.openshift.io` resource is:

```go
type Insights struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata"`

	// spec is the specification of the desired behavior of the Insights
	// +kubebuilder:validation:Required
	// +required
	Spec InsightsSpec `json:"spec"`

	// status is the most recently observed status of the Insights operator
	// +optional
	Status InsightsStatus `json:"status"`
}

type InsightsSpec struct {
	OperatorSpec `json:",inline"`
}

> type InsightsStatus struct {

Reviewer 2: please embed `operatorv1.OperatorStatus`

Author: Embedded
